### PR TITLE
feat(mpc_lateral_controller): dynamic steer rate lim

### PR DESCRIPTION
**Pull request by takayuki5168**
## Description

For dynamic path planning which is not so stable, the dynamic steer rate limit is applied to each reference path point depending on its curvature.

## Tests performed

Not applicable.

## Effects on system behavior

For dynamic path planning whose shape is not so stable, the dynamic steer rate limitation is applied to each reference path point depending on its curvature.

## Pre-review checklist for the PR author

The PR author **must** check the checkboxes below when creating the PR.

- [x] I've confirmed the [contribution guidelines].
- [x] The PR follows the [pull request guidelines].

## In-review checklist for the PR reviewers

The PR reviewers **must** check the checkboxes below before approval.

- [x] The PR follows the [pull request guidelines].

## Post-review checklist for the PR author

The PR author **must** check the checkboxes below before merging.

- [ ] There are no open discussions or they are tracked via tickets.

After all checkboxes are checked, anyone who has write access can merge the PR.

[contribution guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/
[pull request guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/pull-request-guidelines/

### CODE REVIEW

File: control/mpc_lateral_controller/src/mpc.cpp [hunk 1]
```diff
@@ -666,12 +666,29 @@ bool MPC::executeOptimization(
...
+  VectorXd ubA(DIM_U_N);
+  for (int i = 0; i < DIM_U_N; ++i) {
+    const double adaptive_steer_rate_lim = get_adaptive_steer_rate_lim(traj.smooth_k.at(i));
+    const double adaptive_delta_steer_lim = adaptive_steer_rate_lim * prediction_dt;
+    lbA(i) = -adaptive_delta_steer_lim;
+    ubA(i) = adaptive_delta_steer_lim;
+  }
+  const double adaptive_steer_rate_lim = get_adaptive_steer_rate_lim(traj.smooth_k.at(0));
+  lbA(0, 0) = m_raw_steer_cmd_prev - adaptive_steer_rate_lim * m_ctrl_period;
+  ubA(0, 0) = m_raw_steer_cmd_prev + adaptive_steer_rate_lim * m_ctrl_period;
```
Comment (takayuki5168) [inline control/mpc_lateral_controller/src/mpc.cpp:691]: I don't understand what this part meant originally.

File: control/mpc_lateral_controller/param/lateral_controller_defaults.param.yaml [hunk 2]
```diff
@@ -47,7 +47,10 @@
     vehicle_model_type: "kinematics" # vehicle model type for mpc prediction. option is kinematics, kinematics_no_delay, and dynamics
     input_delay: 0.24 # steering input delay time for delay compensation
     vehicle_model_steer_tau: 0.3 # steering dynamics time constant (1d approximation) [s]
-    steer_rate_lim_dps: 600.0 # steering angle rate limit [deg/s]
+    steer_rate_lim_dps_list_by_curvature: [10.0, 20.0]            # steering angle rate limit list depending on curvature [deg/s]
+    curvature_list_for_steer_rate_lim: [0.001, 0.01]              # curvature list for steering angle rate limit interpolation [/m]
+    steer_rate_lim_dps_list_by_velocity: [10.0, 20.0]             # steering angle rate limit list depending on velocity [deg/s]
```
Comment (TakaHoribe) [inline control/mpc_lateral_controller/param/lateral_controller_defaults.param.yaml:52]: Should `steer_rate_lim_dps_list` be in the opposite order?  It should be `[20.0, 10.0]`, not `[10.0, 20.0]`

File: control/mpc_lateral_controller/test/test_mpc.cpp [hunk 3]
```diff
@@ -168,7 +168,8 @@ class MPCTest : public ::testing::Test
     mpc.m_admissible_position_error = admissible_position_error;
     mpc.m_admissible_yaw_error_rad = admissible_yaw_error_rad;
     mpc.m_steer_lim = steer_lim;
-    mpc.m_steer_rate_lim = steer_rate_lim;
+    mpc.m_steer_rate_lim_map_by_curvature.push_back({0.0, steer_rate_lim});
+    mpc.m_steer_rate_lim_map_by_velocity.push_back({0.0, steer_rate_lim});
```
Comment (TakaHoribe) [inline control/mpc_lateral_controller/test/test_mpc.cpp:172]: mino thoughr: use `emplace_back()` when you call the default constructor.

```suggestion
    mpc.m_steer_rate_lim_map_by_velocity.emplace_back(0.0, steer_rate_lim);
```

File: control/mpc_lateral_controller/src/mpc_lateral_controller.cpp [hunk 4]
```diff
@@ -204,17 +223,6 @@ trajectory_follower::LateralOutput MpcLateralController::run(
...
-
-  // if the vehicle is stopped, set steering angle rate limit to a large value to get a proper
-  // steering value from mpc.
-  // TODO(someone): solve mpc cannot create output in low steering rate limit problem
-  if (is_vehicle_stopped) {
-    m_mpc.m_steer_rate_lim = std::numeric_limits<double>::max();
-  } else {
-    m_mpc.m_steer_rate_lim = m_steer_rate_lim;
-  }
-
```
Comment (takayuki5168) [inline control/mpc_lateral_controller/src/mpc_lateral_controller.cpp:217]: This was moved in mpc.cpp

File: control/mpc_lateral_controller/test/test_mpc.cpp (hunk 3 above)
Comment (takayuki5168) [reply control/mpc_lateral_controller/test/test_mpc.cpp]: DONE

**Review by TakaHoribe — APPROVED** [review]
LGTM

**Pull request merged**